pages/blog/posts/understanding-lexical-dynamic-scopes.md

```diff
@@ -0,0 +1,252 @@
+instance with a single object property `bar`. On the right, the evaluation path
+on the JSON Schema tree representation for validating the JSON instance,
+following schema references accordingly:
+
+![A tree representation of the JSON Schema evaluation
+path](/img/posts/2024/understanding-lexical-dynamic-scopes/evaluation-path.webp)
+
+In comparison to the lexical scope, the dynamic scope of a schema cannot always
+be statically determined. For example, for schemas that make use of logic
+applicator keywords such as `if` or `oneOf`, the evaluation path may vary
+depending on the instance. In the previous example, the evaluation path would
+also vary if the instance document declared a `foo` property.
+
+### Lexical and Dynamic Scopes Sometimes Align
+
+When leaving references aside, schema evaluation is always top down. For
+example, consider a schema that defines subschemas through the use of
+applicator keywords. First, the schema is evaluated, and only then its
+subschemas are considered:
+
+![A visual representation of top down schema evaluation](/img/posts/2024/understanding-lexical-dynamic-scopes/evaluation-top-down.webp)
+
+In the simple case where a schema is (1) evaluated from the top level, (2)
+consists of a single schema resource, and (3) does not make use of reference
+keywords (like `$ref` or `$dynamicRef`), the lexical and dynamic scopes of its
+subschemas are the same. 
+
+In the following example, the lexical and dynamic scopes for the subschema
+highlighted in red align:
+
+![A simple example where lexical and dynamic scopes align](/img/posts/2024/understanding-lexical-dynamic-scopes/aligned.webp)
+
+### References and the Evaluation Path
+
+References are the only exception to top down evaluation. By definition,
+references (both static and dynamic) make the evaluation process break from
+this mode by *jumping* to another location (as we saw in the evaluation path
+figure at the beginning of the Dynamic Scope section). As a consequence, when
+the evaluation process encounters a reference, lexical scopes and dynamic
+scopes will typically diverge.
+
+To recap, for the lexical scope, following a reference involves abandoning the
+lexical scope of the reference schema and entering the lexical scope of the
+destination schema. In comparison, for the dynamic scope, <i>following a
+reference involves retaining the current dynamic scope, and adding the
+destination schema to it</i>.
+
+Consider the following example of a schema that follows a reference to an
+anchor. On the left, the lexical scope of the destination schema (highlighted
+in red), consists of it parent schema (highlighted in green), and its
+grandparent schema (highlighted in blue). However, on the right, the dynamic
+scope of the destination schema (highlighted in red) consists of the schema
+that referenced the anchor (highlighted in green) and the schema that led to
+the evaluation of such reference (highlighted in blue):
```

Comment:
This isn't correct. The schema highlighted in green is not part of the lexical scope. This example is one where the lexical and dynamic scopes should align.